syntax = "proto3";

package aea.eightballer.user_interaction.v0_1_0;

message UserInteractionMessage{

  // Custom Types
  message ErrorCode{
    enum ErrorCodeEnum {
      UNKNOWN_ERROR = 0;
      COMMAND_NOT_SUPPORTED = 1;
      COMMAND_FAILED = 2;
      NOTIFICATION_FAILED = 3;
      INVALID_DATA = 4;
    }
    ErrorCodeEnum error_code = 1;
  }


  // Performatives and contents
  message Notification_Performative{
    string body = 1;
    string title = 2;
    bool title_is_set = 3;
    string attach = 4;
    bool attach_is_set = 5;
  }

  message User_Command_Performative{
    string command = 1;
    map<string, bytes> data = 2;
    bool data_is_set = 3;
  }

  message Request_Confirmation_Performative{
    string message = 1;
  }

  message User_Confirmation_Performative{
    bool confirmed = 1;
  }

  message Error_Performative{
    ErrorCode error_code = 1;
    string error_msg = 2;
    map<string, bytes> error_data = 3;
    bool error_data_is_set = 4;
  }

  message End_Performative{
  }


  oneof performative{
    End_Performative end = 5;
    Error_Performative error = 6;
    Notification_Performative notification = 7;
    Request_Confirmation_Performative request_confirmation = 8;
    User_Command_Performative user_command = 9;
    User_Confirmation_Performative user_confirmation = 10;
  }
}
